### PR TITLE
Fix open redirect check for more cases

### PR DESCRIPTION
**Pull request by lafriks**
If redirect_to parameter has set value starting with `\\example.com`  redirect will be created with header `Location: /\\example.com` that will redirect to example.com domain.

### CODE REVIEW

**Review by silverwind — APPROVED** [review]

File: modules/context/context_response.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func (ctx *Context) RedirectToFirst(location ...string) {
 
 		// Unfortunately browsers consider a redirect Location with preceding "//" and "/\" as meaning redirect to "http(s)://REST_OF_PATH"
```
Comment (silverwind) [inline modules/context/context_response.go]: Comment is not accurate anymore

**Comment by silverwind** [comment]
How about a regex like `[\\/]{2}`?

File: modules/context/context_response.go (hunk 1 above)
Comment (lafriks) [reply modules/context/context_response.go]: comment updated

**Comment by lafriks** [comment]
I would avoid regex for this use case

**Review by silverwind — APPROVED** [review]

File: modules/context/context_response.go [hunk 2]
```diff
@@ -49,9 +49,9 @@ func (ctx *Context) RedirectToFirst(location ...string) {
 			continue
 		}
 
-		// Unfortunately browsers consider a redirect Location with preceding "//" and "/\" as meaning redirect to "http(s)://REST_OF_PATH"
+		// Unfortunately browsers consider a redirect Location with preceding "//", "\\" and "/\" as meaning redirect to "http(s)://REST_OF_PATH"
 		// Therefore we should ignore these redirect locations to prevent open redirects
-		if len(loc) > 1 && loc[0] == '/' && (loc[1] == '/' || loc[1] == '\\') {
+		if len(loc) > 1 && (loc[0] == '/' || loc[0] == '\\') && (loc[1] == '/' || loc[1] == '\\') {
```
Comment (delvh) [inline modules/context/context_response.go:54]: Your code also does not allow `"\/"`. I guess that's expected?

**Review by delvh — APPROVED** [review]

File: modules/context/context_response.go (hunk 2 above)
Comment (lafriks) [reply modules/context/context_response.go:54]: yes, while this one probably is not dangerous I see no real use case for it anyway

**Comment by lafriks** [comment]
As this part (context package) of code has no unit tests that would would allow creating context etc to test on this is not easy one to add one, especially for such small change

**Comment by lafriks** [comment]
This still is not perfect solution as this would need to be actually context aware (like redirecting only to valid Gitea URLs etc) like stated in this FIXME comment:
https://github.com/go-gitea/gitea/blob/263ed09bbfebfbd7f53f2165c2175c2afab5a42d/routers/web/user/profile.go#L366

In future this should be fixed so I would avoid making changes just to be able to test small portion of this

**Pull request merged**